### ISSUE TITLE
Any experience with Nissan Leaf EV?

### ISSUE THREAD

**Issue by TommySharpNZ** [meatpiHQ/wican-fw#51]
Has anyone had experience using the MeatPi in a Nissan Leaf?
I''m trying to find a way to get things like Battery SoC into Home Assistant...

**Comment by eludias** [comment]
Hi @TommySharpNZ,

I'm on the same journey, but for a Chevy Bolt instead of the Nissan Leaf: to use automation to optimize charging.

Expect to spend time on this though, since MeatPI still has rough edges. To get started:
- Buy Torque Pro, Engine Link or Car Scanner Pro; these apps support custom PIDs
- Set MeatPI to ELM327 protocol on Wifi.
- Search for a Nissan Leaf specific CSV file with the proprietary PIDs. For example, [OBD-PIDs-for-HKMC-EVs](https://github.com/JejuSoul/OBD-PIDs-for-HKMC-EVs) has them for other EVs.

This will learn you:
- That MeatPI works;
- What CAN PID to get out of the Leaf;
- How to interpret this data

With this knowledge you can enable MQTT and tailer the Javascript in Home Assistent's Red Node.

Good luck!

**Comment by meatpiHQ** [comment]
@TommySharpNZ You should be able to read the SoC. You need to find the right pid request for your car.

Good place to start:

https://github.com/meatpiHQ/wican-fw/issues/17#issuecomment-1472276657
https://github.com/meatpiHQ/wican-fw/discussions/23

**Comment by robwolff3** [comment]
Hey @eludias , I'm also using the MeatPi with a Chevy Bolt. Can you share what you've done so far?

Would be much appreciated :)

**Comment by robwolff3** [comment]
For anyone else doing this you can find the Chevy Bolt PID's here: https://docs.google.com/spreadsheets/d/1sY5n8WFu52U6a4_mg3MdcGcmDk3scAP8_muSy-BlXPc/edit?usp=sharing

Getting Torque Pro to work was easy using the instructions here https://allev.info/boltpids/ but I was getting stuck on trying to translate the ModeAndPID to work with the [meatPI Home Assistant example](https://github.com/meatpiHQ/wican-fw#home-assistant)

I eventually figured it out after reading https://github.com/meatpiHQ/wican-fw/issues/17#issuecomment-1641850205 and translating it for the bolt.

Here is an example of Battery Coolant Tempature - ModeAndPID: 2241A4:
```
{
  "bus": "0",
  "type": "tx",
  "frame": [
    {
      "id": 2020,
      "dlc": 8,
      "rtr": false,
      "extd": false,
      "data": [
        3,
        34,
        65,
        164,
        170,
        170,
        170,
        170
      ]
    }
  ]
}
```
Converting Hex to Decimal:
- id, I used the OBD Header from Torque Pro: 0x7E4 = 2020
- "3 because the pid is 4 bytes long"
- Mode is the first two characters of the ModeAndPID: 0x22 = 34
- The PID is the next 4 characters: 0x41 and 0xA4 = 65 and 164

Then I received the response with:
```
var obj = Object();

for (let i = 0; i < msg.payload.frame.length; i++) {
    //check frame id
    if (msg.payload.frame[0].id == 2028 && msg.payload.frame[0].data[2] == 65 && msg.payload.frame[0].data[3] == 164) {
        obj.amb_temp = msg.payload.frame[0].data[4] - 40;
        msg.payload = obj;
        return msg;
    }
}
```

**Comment by camueller** [comment]
> Has anyone had experience using the MeatPi in a Nissan Leaf? I''m trying to find a way to get things like Battery SoC into Home Assistant...

Yes, I've got it working: https://github.com/meatpiHQ/wican-fw/discussions/93#discussioncomment-8583799
